// wechaty/dist/cjs/src/user-modules/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UrlLinkImpl = exports.TagImpl = exports.DelayImpl = exports.RoomInvitationImpl = exports.RoomImpl = exports.PostImpl = exports.MoneyImpl = exports.MomentImpl = exports.MiniProgramImpl = exports.MessageImpl = exports.LocationImpl = exports.ImageImpl = exports.FriendshipImpl = exports.FavoriteImpl = exports.ContactSelfImpl = exports.ContactImpl = exports.wechatifyUserModule = void 0;
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
const contact_js_1 = require("./contact.js");
Object.defineProperty(exports, "ContactImpl", { enumerable: true, get: function () { return contact_js_1.ContactImpl; } });
const contact_self_js_1 = require("./contact-self.js");
Object.defineProperty(exports, "ContactSelfImpl", { enumerable: true, get: function () { return contact_self_js_1.ContactSelfImpl; } });
const favorite_js_1 = require("./favorite.js");
Object.defineProperty(exports, "FavoriteImpl", { enumerable: true, get: function () { return favorite_js_1.FavoriteImpl; } });
const friendship_js_1 = require("./friendship.js");
Object.defineProperty(exports, "FriendshipImpl", { enumerable: true, get: function () { return friendship_js_1.FriendshipImpl; } });
const image_js_1 = require("./image.js");
Object.defineProperty(exports, "ImageImpl", { enumerable: true, get: function () { return image_js_1.ImageImpl; } });
const location_js_1 = require("./location.js");
Object.defineProperty(exports, "LocationImpl", { enumerable: true, get: function () { return location_js_1.LocationImpl; } });
const message_js_1 = require("./message.js");
Object.defineProperty(exports, "MessageImpl", { enumerable: true, get: function () { return message_js_1.MessageImpl; } });
const mini_program_js_1 = require("./mini-program.js");
Object.defineProperty(exports, "MiniProgramImpl", { enumerable: true, get: function () { return mini_program_js_1.MiniProgramImpl; } });
const moment_js_1 = require("./moment.js");
Object.defineProperty(exports, "MomentImpl", { enumerable: true, get: function () { return moment_js_1.MomentImpl; } });
const money_js_1 = require("./money.js");
Object.defineProperty(exports, "MoneyImpl", { enumerable: true, get: function () { return money_js_1.MoneyImpl; } });
const post_js_1 = require("./post.js");
Object.defineProperty(exports, "PostImpl", { enumerable: true, get: function () { return post_js_1.PostImpl; } });
const room_js_1 = require("./room.js");
Object.defineProperty(exports, "RoomImpl", { enumerable: true, get: function () { return room_js_1.RoomImpl; } });
const room_invitation_js_1 = require("./room-invitation.js");
Object.defineProperty(exports, "RoomInvitationImpl", { enumerable: true, get: function () { return room_invitation_js_1.RoomInvitationImpl; } });
const tag_js_1 = require("./tag.js");
Object.defineProperty(exports, "TagImpl", { enumerable: true, get: function () { return tag_js_1.TagImpl; } });
const url_link_js_1 = require("./url-link.js");
Object.defineProperty(exports, "UrlLinkImpl", { enumerable: true, get: function () { return url_link_js_1.UrlLinkImpl; } });
const delay_js_1 = require("./delay.js");
Object.defineProperty(exports, "DelayImpl", { enumerable: true, get: function () { return delay_js_1.DelayImpl; } });
const wechatify_js_1 = require("../user-mixins/wechatify.js");
Object.defineProperty(exports, "wechatifyUserModule", { enumerable: true, get: function () { return wechatify_js_1.wechatifyUserModule; } });
