// wechaty-puppet/dist/cjs/src/schemas/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YOU = exports.ScanStatus = exports.PUPPET_EVENT_DICT = exports.PayloadType = exports.MessageType = exports.ImageType = exports.FriendshipType = exports.FriendshipSceneType = exports.ContactType = exports.ContactGender = exports.CHAT_EVENT_DICT = void 0;
const contact_js_1 = require("./contact.js");
Object.defineProperty(exports, "ContactGender", { enumerable: true, get: function () { return contact_js_1.ContactGender; } });
Object.defineProperty(exports, "ContactType", { enumerable: true, get: function () { return contact_js_1.ContactType; } });
const event_js_1 = require("./event.js");
Object.defineProperty(exports, "ScanStatus", { enumerable: true, get: function () { return event_js_1.ScanStatus; } });
const friendship_js_1 = require("./friendship.js");
Object.defineProperty(exports, "FriendshipType", { enumerable: true, get: function () { return friendship_js_1.FriendshipType; } });
Object.defineProperty(exports, "FriendshipSceneType", { enumerable: true, get: function () { return friendship_js_1.FriendshipSceneType; } });
const image_js_1 = require("./image.js");
Object.defineProperty(exports, "ImageType", { enumerable: true, get: function () { return image_js_1.ImageType; } });
const message_js_1 = require("./message.js");
Object.defineProperty(exports, "MessageType", { enumerable: true, get: function () { return message_js_1.MessageType; } });
const payload_js_1 = require("./payload.js");
Object.defineProperty(exports, "PayloadType", { enumerable: true, get: function () { return payload_js_1.PayloadType; } });
const puppet_js_1 = require("./puppet.js");
Object.defineProperty(exports, "CHAT_EVENT_DICT", { enumerable: true, get: function () { return puppet_js_1.CHAT_EVENT_DICT; } });
Object.defineProperty(exports, "PUPPET_EVENT_DICT", { enumerable: true, get: function () { return puppet_js_1.PUPPET_EVENT_DICT; } });
Object.defineProperty(exports, "YOU", { enumerable: true, get: function () { return puppet_js_1.YOU; } });
