// wechaty/dist/cjs/src/mods/impls.js
"use strict";
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WechatyImpl = exports.UrlLinkImpl = exports.TagImpl = exports.DelayImpl = exports.RoomInvitationImpl = exports.RoomImpl = exports.PostImpl = exports.MoneyImpl = exports.MomentImpl = exports.MiniProgramImpl = exports.MessageImpl = exports.LocationImpl = exports.ImageImpl = exports.FriendshipImpl = exports.FavoriteImpl = exports.ContactSelfImpl = exports.ContactImpl = void 0;
/**
 * Huan(202111): We export all names with specific postfix to make it easier to be recongnized
 *  with less misunderstandings.
 *
 * Postfixes:
 *  - Interface:    type, the `interface
 *  - Constructor:  type, the `Constructor<interface>`
 *  - Impl:         value, the `class`
 *
 *  - no postfix:  value, the `class`
 */
var mod_js_1 = require("../user-modules/mod.js");
Object.defineProperty(exports, "ContactImpl", { enumerable: true, get: function () { return mod_js_1.ContactImpl; } });
Object.defineProperty(exports, "ContactSelfImpl", { enumerable: true, get: function () { return mod_js_1.ContactSelfImpl; } });
Object.defineProperty(exports, "FavoriteImpl", { enumerable: true, get: function () { return mod_js_1.FavoriteImpl; } });
Object.defineProperty(exports, "FriendshipImpl", { enumerable: true, get: function () { return mod_js_1.FriendshipImpl; } });
Object.defineProperty(exports, "ImageImpl", { enumerable: true, get: function () { return mod_js_1.ImageImpl; } });
Object.defineProperty(exports, "LocationImpl", { enumerable: true, get: function () { return mod_js_1.LocationImpl; } });
Object.defineProperty(exports, "MessageImpl", { enumerable: true, get: function () { return mod_js_1.MessageImpl; } });
Object.defineProperty(exports, "MiniProgramImpl", { enumerable: true, get: function () { return mod_js_1.MiniProgramImpl; } });
Object.defineProperty(exports, "MomentImpl", { enumerable: true, get: function () { return mod_js_1.MomentImpl; } });
Object.defineProperty(exports, "MoneyImpl", { enumerable: true, get: function () { return mod_js_1.MoneyImpl; } });
Object.defineProperty(exports, "PostImpl", { enumerable: true, get: function () { return mod_js_1.PostImpl; } });
Object.defineProperty(exports, "RoomImpl", { enumerable: true, get: function () { return mod_js_1.RoomImpl; } });
Object.defineProperty(exports, "RoomInvitationImpl", { enumerable: true, get: function () { return mod_js_1.RoomInvitationImpl; } });
Object.defineProperty(exports, "DelayImpl", { enumerable: true, get: function () { return mod_js_1.DelayImpl; } });
Object.defineProperty(exports, "TagImpl", { enumerable: true, get: function () { return mod_js_1.TagImpl; } });
Object.defineProperty(exports, "UrlLinkImpl", { enumerable: true, get: function () { return mod_js_1.UrlLinkImpl; } });
var mod_js_2 = require("../wechaty/mod.js");
Object.defineProperty(exports, "WechatyImpl", { enumerable: true, get: function () { return mod_js_2.WechatyImpl; } });
