// wechaty-puppet/dist/cjs/src/mixins/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readyMixin = exports.validateMixin = exports.tagMixin = exports.serviceMixin = exports.roomMixin = exports.roomMemberMixin = exports.roomInvitationMixin = exports.miscMixin = exports.messageMixin = exports.memoryMixin = exports.loginMixin = exports.friendshipMixin = exports.contactMixin = exports.cacheMixin = void 0;
const cache_mixin_js_1 = require("./cache-mixin.js");
Object.defineProperty(exports, "cacheMixin", { enumerable: true, get: function () { return cache_mixin_js_1.cacheMixin; } });
const contact_mixin_js_1 = require("./contact-mixin.js");
Object.defineProperty(exports, "contactMixin", { enumerable: true, get: function () { return contact_mixin_js_1.contactMixin; } });
const friendship_mixin_js_1 = require("./friendship-mixin.js");
Object.defineProperty(exports, "friendshipMixin", { enumerable: true, get: function () { return friendship_mixin_js_1.friendshipMixin; } });
const login_mixin_js_1 = require("./login-mixin.js");
Object.defineProperty(exports, "loginMixin", { enumerable: true, get: function () { return login_mixin_js_1.loginMixin; } });
const memory_mixin_js_1 = require("./memory-mixin.js");
Object.defineProperty(exports, "memoryMixin", { enumerable: true, get: function () { return memory_mixin_js_1.memoryMixin; } });
const message_mixin_js_1 = require("./message-mixin.js");
Object.defineProperty(exports, "messageMixin", { enumerable: true, get: function () { return message_mixin_js_1.messageMixin; } });
const misc_mixin_js_1 = require("./misc-mixin.js");
Object.defineProperty(exports, "miscMixin", { enumerable: true, get: function () { return misc_mixin_js_1.miscMixin; } });
const ready_mixin_js_1 = require("./ready-mixin.js");
Object.defineProperty(exports, "readyMixin", { enumerable: true, get: function () { return ready_mixin_js_1.readyMixin; } });
const room_invitation_mixin_js_1 = require("./room-invitation-mixin.js");
Object.defineProperty(exports, "roomInvitationMixin", { enumerable: true, get: function () { return room_invitation_mixin_js_1.roomInvitationMixin; } });
const room_member_mixin_js_1 = require("./room-member-mixin.js");
Object.defineProperty(exports, "roomMemberMixin", { enumerable: true, get: function () { return room_member_mixin_js_1.roomMemberMixin; } });
const room_mixin_js_1 = require("./room-mixin.js");
Object.defineProperty(exports, "roomMixin", { enumerable: true, get: function () { return room_mixin_js_1.roomMixin; } });
const service_mixin_js_1 = require("./service-mixin.js");
Object.defineProperty(exports, "serviceMixin", { enumerable: true, get: function () { return service_mixin_js_1.serviceMixin; } });
const tag_mixin_js_1 = require("./tag-mixin.js");
Object.defineProperty(exports, "tagMixin", { enumerable: true, get: function () { return tag_mixin_js_1.tagMixin; } });
const validate_mixin_js_1 = require("./validate-mixin.js");
Object.defineProperty(exports, "validateMixin", { enumerable: true, get: function () { return validate_mixin_js_1.validateMixin; } });
