// wechaty-puppet-service/dist/cjs/src/auth/env-vars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_NAME = exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_KEY = exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_CERT = exports.WECHATY_PUPPET_SERVICE_TLS_CA_CERT = exports.WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_SERVER = exports.WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_CLIENT = void 0;
/**
 * Environment variables containing newlines in Node?
 *  `replace(/\\n/g, '\n')`
 *    https://stackoverflow.com/a/36439803/1123955
 */
const WECHATY_PUPPET_SERVICE_TLS_CA_CERT = (v) => v
    || process.env['WECHATY_PUPPET_SERVICE_TLS_CA_CERT']?.replace(/\\n/g, '\n');
exports.WECHATY_PUPPET_SERVICE_TLS_CA_CERT = WECHATY_PUPPET_SERVICE_TLS_CA_CERT;
const WECHATY_PUPPET_SERVICE_TLS_SERVER_CERT = (v) => v
    || process.env['WECHATY_PUPPET_SERVICE_TLS_SERVER_CERT']?.replace(/\\n/g, '\n');
exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_CERT = WECHATY_PUPPET_SERVICE_TLS_SERVER_CERT;
const WECHATY_PUPPET_SERVICE_TLS_SERVER_KEY = (v) => v
    || process.env['WECHATY_PUPPET_SERVICE_TLS_SERVER_KEY']?.replace(/\\n/g, '\n');
exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_KEY = WECHATY_PUPPET_SERVICE_TLS_SERVER_KEY;
const WECHATY_PUPPET_SERVICE_TLS_SERVER_NAME = (v) => v
    || process.env['WECHATY_PUPPET_SERVICE_TLS_SERVER_NAME']?.replace(/\\n/g, '\n');
exports.WECHATY_PUPPET_SERVICE_TLS_SERVER_NAME = WECHATY_PUPPET_SERVICE_TLS_SERVER_NAME;
const WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_SERVER = (v) => typeof v === 'undefined'
    ? process.env['WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_SERVER'] === 'true'
    : v;
exports.WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_SERVER = WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_SERVER;
const WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_CLIENT = (v) => typeof v === 'undefined'
    ? process.env['WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_CLIENT'] === 'true'
    : v;
exports.WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_CLIENT = WECHATY_PUPPET_SERVICE_NO_TLS_INSECURE_CLIENT;
