// wechaty-puppet/dist/esm/src/mixins/mod.js
import { cacheMixin, } from './cache-mixin.js';
import { contactMixin, } from './contact-mixin.js';
import { friendshipMixin, } from './friendship-mixin.js';
import { loginMixin, } from './login-mixin.js';
import { memoryMixin, } from './memory-mixin.js';
import { messageMixin, } from './message-mixin.js';
import { miscMixin } from './misc-mixin.js';
import { readyMixin, } from './ready-mixin.js';
import { roomInvitationMixin, } from './room-invitation-mixin.js';
import { roomMemberMixin, } from './room-member-mixin.js';
import { roomMixin, } from './room-mixin.js';
import { serviceMixin, } from './service-mixin.js';
import { tagMixin, } from './tag-mixin.js';
import { tapMixin, } from './tap-mixin.js';
import { postMixin, } from './post-mixin.js';
import { sayableMixin, } from '../mixins/sayable-mixin.js';
import { validateMixin, } from './validate-mixin.js';
export { cacheMixin, contactMixin, friendshipMixin, loginMixin, memoryMixin, messageMixin, miscMixin, postMixin, readyMixin, roomInvitationMixin, roomMemberMixin, roomMixin, serviceMixin, tagMixin, tapMixin, sayableMixin, validateMixin, };
