// wechaty/dist/cjs/src/pure-functions/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unifyEmoji = exports.unescapeHtml = exports.stripHtml = exports.stripEmoji = exports.plainText = exports.digestEmoji = exports.retryPolicy = void 0;
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
var retry_policy_js_1 = require("./retry-policy.js");
Object.defineProperty(exports, "retryPolicy", { enumerable: true, get: function () { return retry_policy_js_1.retryPolicy; } });
var xml_js_1 = require("./xml.js");
Object.defineProperty(exports, "digestEmoji", { enumerable: true, get: function () { return xml_js_1.digestEmoji; } });
Object.defineProperty(exports, "plainText", { enumerable: true, get: function () { return xml_js_1.plainText; } });
Object.defineProperty(exports, "stripEmoji", { enumerable: true, get: function () { return xml_js_1.stripEmoji; } });
Object.defineProperty(exports, "stripHtml", { enumerable: true, get: function () { return xml_js_1.stripHtml; } });
Object.defineProperty(exports, "unescapeHtml", { enumerable: true, get: function () { return xml_js_1.unescapeHtml; } });
Object.defineProperty(exports, "unifyEmoji", { enumerable: true, get: function () { return xml_js_1.unifyEmoji; } });
