// cmd-ts/dist/esm/multiflag.js
import { findOption } from './newparser/findOption';
import { boolean } from './flag';
import * as Result from './Result';
/**
 * Like `option`, but can accept multiple options, and expects a decoder from a list of strings.
 * An error will highlight all option occurences.
 */
export function multiflag(config) {
    return {
        helpTopics() {
            var _a;
            let usage = `--${config.long}`;
            if (config.short) {
                usage += `, -${config.short}`;
            }
            return [
                {
                    category: 'flags',
                    usage,
                    defaults: [],
                    description: (_a = config.description) !== null && _a !== void 0 ? _a : 'self explanatory',
                },
            ];
        },
        register(opts) {
            opts.forceFlagLongNames.add(config.long);
            if (config.short) {
                opts.forceFlagShortNames.add(config.short);
            }
        },
        async parse({ nodes, visitedNodes, }) {
            var _a, _b;
            const options = findOption(nodes, {
                longNames: [config.long],
                shortNames: config.short ? [config.short] : [],
            }).filter(x => !visitedNodes.has(x));
            for (const option of options) {
                visitedNodes.add(option);
            }
            const optionValues = [];
            const errors = [];
            for (const option of options) {
                const decoded = await Result.safeAsync(boolean.from((_b = (_a = option.value) === null || _a === void 0 ? void 0 : _a.node.raw) !== null && _b !== void 0 ? _b : 'true'));
                if (Result.isErr(decoded)) {
                    errors.push({ nodes: [option], message: decoded.error.message });
                }
                else {
                    optionValues.push(decoded.value);
                }
            }
            if (errors.length > 0) {
                return Result.err({
                    errors,
                });
            }
            const multiDecoded = await Result.safeAsync(config.type.from(optionValues));
            if (Result.isErr(multiDecoded)) {
                return Result.err({
                    errors: [
                        {
                            nodes: options,
                            message: multiDecoded.error.message,
                        },
                    ],
                });
            }
            return multiDecoded;
        },
    };
}
