// wechaty/dist/esm/src/user-modules/mini-program.js
import { log } from '../config.js';
import { validationMixin } from '../user-mixins/validation.js';
import { wechatifyMixinBase, } from '../user-mixins/wechatify.js';
class MiniProgramMixin extends wechatifyMixinBase() {
    payload;
    /**
     *
     * Create
     *
     */
    static async create() {
        log.verbose('MiniProgram', 'create()');
        // TODO: get appid and username from wechat
        const payload = {
            appid: 'todo',
            description: 'todo',
            pagePath: 'todo',
            thumbKey: 'todo',
            thumbUrl: 'todo',
            title: 'todo',
            username: 'todo',
        };
        return new this(payload);
    }
    /*
     * @hideconstructor
     */
    constructor(payload) {
        super();
        this.payload = payload;
        log.verbose('MiniProgram', 'constructor()');
        // Huan(202110): it is ok to create a raw one without wechaty instance
        // guardWechatifyClass.call(this, MiniProgram)
    }
    appid() {
        return this.payload.appid;
    }
    title() {
        return this.payload.title;
    }
    pagePath() {
        return this.payload.pagePath;
    }
    username() {
        return this.payload.username;
    }
    description() {
        return this.payload.description;
    }
    thumbUrl() {
        return this.payload.thumbUrl;
    }
    thumbKey() {
        return this.payload.thumbKey;
    }
}
class MiniProgramImpl extends validationMixin(MiniProgramMixin)() {
}
export { MiniProgramImpl, };
