// wechaty-token/dist/cjs/src/grpc-js.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uriToString = exports.resolverManager = exports.parseUri = exports.Metadata = exports.GrpcStatus = exports.BackoffTimeout = void 0;
const uri_parser_js_1 = require("@grpc/grpc-js/build/src/uri-parser.js");
Object.defineProperty(exports, "parseUri", { enumerable: true, get: function () { return uri_parser_js_1.parseUri; } });
Object.defineProperty(exports, "uriToString", { enumerable: true, get: function () { return uri_parser_js_1.uriToString; } });
const resolverManager = __importStar(require("@grpc/grpc-js/build/src/resolver.js"));
exports.resolverManager = resolverManager;
const metadata_js_1 = require("@grpc/grpc-js/build/src/metadata.js");
Object.defineProperty(exports, "Metadata", { enumerable: true, get: function () { return metadata_js_1.Metadata; } });
const backoff_timeout_js_1 = require("@grpc/grpc-js/build/src/backoff-timeout.js");
Object.defineProperty(exports, "BackoffTimeout", { enumerable: true, get: function () { return backoff_timeout_js_1.BackoffTimeout; } });
const grpc_js_1 = require("@grpc/grpc-js");
Object.defineProperty(exports, "GrpcStatus", { enumerable: true, get: function () { return grpc_js_1.status; } });
