// gerror/dist/cjs/src/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapAsyncError = exports.VERSION = exports.TimeoutPromiseGError = exports.timeoutPromise = exports.isGrpcStatus = exports.isGError = exports.isEcmaError = exports.GError = void 0;
const version_js_1 = require("./version.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_js_1.VERSION; } });
const gerror_js_1 = require("./gerror/gerror.js");
Object.defineProperty(exports, "isEcmaError", { enumerable: true, get: function () { return gerror_js_1.isEcmaError; } });
Object.defineProperty(exports, "isGrpcStatus", { enumerable: true, get: function () { return gerror_js_1.isGrpcStatus; } });
Object.defineProperty(exports, "isGError", { enumerable: true, get: function () { return gerror_js_1.isGError; } });
Object.defineProperty(exports, "GError", { enumerable: true, get: function () { return gerror_js_1.GError; } });
const mod_js_1 = require("./timeout-promise/mod.js");
Object.defineProperty(exports, "timeoutPromise", { enumerable: true, get: function () { return mod_js_1.timeoutPromise; } });
Object.defineProperty(exports, "TimeoutPromiseGError", { enumerable: true, get: function () { return mod_js_1.TimeoutPromiseGError; } });
const wrap_async_error_js_1 = require("./wrap-async-error/wrap-async-error.js");
Object.defineProperty(exports, "wrapAsyncError", { enumerable: true, get: function () { return wrap_async_error_js_1.wrapAsyncError; } });
