// wechaty-token/dist/cjs/bin/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * https://github.com/wechaty/token
 *
 * Author: Huan <zixia@zixia.net>
 * License: Apache-2.0
 *
 * CLI Apps in TypeScript with `cmd-ts` (Part 1)
 *  Using `cmd-ts` to easily build a type-safe TypeScript CLI app
 *
 *  https://gal.hagever.com/posts/type-safe-cli-apps-in-typescript-with-cmd-ts-part-1/
 */
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
const version_js_1 = require("../src/version.js");
const generate_js_1 = require("../src/cli/generate.js");
const discover_js_1 = require("../src/cli/discover.js");
const wechatyToken = (0, cmd_ts_1.subcommands)({
    name: 'wechaty-token',
    description: 'Wechaty utility for discovering and generating tokens',
    version: version_js_1.VERSION,
    cmds: {
        generate: generate_js_1.generate,
        discover: discover_js_1.discover,
    },
});
(0, cmd_ts_1.run)((0, cmd_ts_1.binary)(wechatyToken), process.argv).catch(console.error);
