// wechaty/dist/cjs/src/mods/mod.js
"use strict";
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoClient = exports.VERSION = exports.qrcodeValueToImageUrl = exports.config = exports.log = exports.ScanStatus = exports.helpers = exports.impls = exports.users = exports.payloads = exports.types = exports.WechatyBuilder = void 0;
var wechaty_builder_js_1 = require("../wechaty-builder.js");
Object.defineProperty(exports, "WechatyBuilder", { enumerable: true, get: function () { return wechaty_builder_js_1.WechatyBuilder; } });
var mod_js_1 = require("../sayable/mod.js");
exports.types = __importStar(require("wechaty-puppet/types"));
exports.payloads = __importStar(require("wechaty-puppet/payloads"));
__exportStar(require("./users.js"), exports);
exports.users = __importStar(require("./users.js"));
exports.impls = __importStar(require("./impls.js"));
exports.helpers = __importStar(require("./helpers.js"));
var types_1 = require("wechaty-puppet/types");
Object.defineProperty(exports, "ScanStatus", { enumerable: true, get: function () { return types_1.ScanStatus; } });
var config_js_1 = require("../config.js");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return config_js_1.log; } });
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_js_1.config; } });
Object.defineProperty(exports, "qrcodeValueToImageUrl", { enumerable: true, get: function () { return config_js_1.qrcodeValueToImageUrl; } });
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_js_1.VERSION; } });
var io_client_js_1 = require("../io-client.js");
Object.defineProperty(exports, "IoClient", { enumerable: true, get: function () { return io_client_js_1.IoClient; } });
