// wechaty-puppet/dist/cjs/src/mixins/friendship-mixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.friendshipMixin = void 0;
const config_js_1 = require("../config.js");
const payload_js_1 = require("../schemas/payload.js");
const friendshipMixin = (mixinBase) => {
    class FriendshipMixin extends mixinBase {
        constructor(...args) {
            super(...args);
            config_js_1.log.verbose('PuppetFriendshipMixin', 'constructor()');
        }
        async friendshipSearch(searchQueryFilter) {
            config_js_1.log.verbose('PuppetFriendshipMixin', 'friendshipSearch("%s")', JSON.stringify(searchQueryFilter));
            if (Object.keys(searchQueryFilter).length !== 1) {
                throw new Error('searchQueryFilter should only include one key for query!');
            }
            if (searchQueryFilter.phone) {
                return this.friendshipSearchPhone(searchQueryFilter.phone);
            }
            else if (searchQueryFilter.weixin) {
                return this.friendshipSearchWeixin(searchQueryFilter.weixin);
            }
            throw new Error(`unknown key from searchQueryFilter: ${Object.keys(searchQueryFilter).join('')}`);
        }
        /**
         * Issue #155 - https://github.com/wechaty/puppet/issues/155
         *
         * @protected
         */
        friendshipPayloadCache(friendshipId) {
            config_js_1.log.silly('PuppetFriendshipMixin', 'friendshipPayloadCache(id=%s) @ %s', friendshipId, this);
            if (!friendshipId) {
                throw new Error('no id');
            }
            const cachedPayload = this.cache.friendship.get(friendshipId);
            if (cachedPayload) {
                // log.silly('PuppetFriendshipMixin', 'friendshipPayloadCache(%s) cache HIT', friendshipId)
            }
            else {
                config_js_1.log.silly('PuppetFriendshipMixin', 'friendshipPayloadCache(%s) cache MISS', friendshipId);
            }
            return cachedPayload;
        }
        async friendshipPayload(friendshipId, newPayload) {
            config_js_1.log.verbose('PuppetFriendshipMixin', 'friendshipPayload(%s)', friendshipId, newPayload
                ? ',' + JSON.stringify(newPayload)
                : '');
            if (typeof newPayload === 'object') {
                await this.cache.friendship.set(friendshipId, newPayload);
                return;
            }
            /**
              * 1. Try to get from cache first
              */
            const cachedPayload = this.friendshipPayloadCache(friendshipId);
            if (cachedPayload) {
                return cachedPayload;
            }
            /**
              * 2. Cache not found
              */
            const rawPayload = await this.friendshipRawPayload(friendshipId);
            const payload = await this.friendshipRawPayloadParser(rawPayload);
            this.cache.friendship.set(friendshipId, payload);
            return payload;
        }
        async friendshipPayloadDirty(id) {
            config_js_1.log.verbose('PuppetFriendshipMixin', 'friendshipPayloadDirty(%s)', id);
            await this.__dirtyPayloadAwait(payload_js_1.PayloadType.Friendship, id);
        }
    }
    return FriendshipMixin;
};
exports.friendshipMixin = friendshipMixin;
