// wechaty-puppet/dist/cjs/src/mixins/login-mixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loginMixin = void 0;
const config_js_1 = require("../config.js");
const loginMixin = (mixinBase) => {
    class LoginMixin extends mixinBase {
        /**
         * @internal used by public API `currentUserId`
         */
        __currentUserId;
        /**
         * The current logged in user id.
         */
        get currentUserId() {
            config_js_1.log.silly('PuppetLoginMixin', 'get currentUserId()');
            if (!this.__currentUserId) {
                throw new Error('not logged in, no this.__currentUserId yet.');
            }
            return this.__currentUserId;
        }
        /**
         * Boolean value indicates whether the user is logged in or not.
         */
        get isLoggedIn() {
            return !!this.__currentUserId;
        }
        constructor(...args) {
            super(...args);
            config_js_1.log.verbose('PuppetLoginMixin', 'constructor()');
        }
        async start() {
            config_js_1.log.verbose('PuppetLoginMixin', 'start()');
            await super.start();
        }
        async stop() {
            config_js_1.log.verbose('PuppetLoginMixin', 'stop()');
            await this.logout();
            await super.stop();
        }
        /**
         * Need to be called internally when the puppet is logined.
         * this method will emit a `login` event
         * @internal for puppet internal usage
         */
        login(userId) {
            config_js_1.log.verbose('PuppetLoginMixin', 'login(%s)', userId);
            if (this.__currentUserId) {
                throw new Error('must logout first before login again!');
            }
            this.__currentUserId = userId;
            this.emit('login', { contactId: userId });
        }
        /**
         * Need to be called internally/externally when the puppet need to be logouted
         * this method will emit a `logout` event,
         *
         * Note: must set `this.currentUserId = undefined` in this function.
         */
        async logout(reason = 'logout()') {
            config_js_1.log.verbose('PuppetLoginMixin', 'logout(%s)', reason);
            if (!this.isLoggedIn) {
                config_js_1.log.verbose('PuppetLoginMixin', 'logout() isLoggedIn === false, do nothing');
                return;
            }
            this.emit('logout', {
                contactId: this.currentUserId,
                data: reason,
            });
            /**
             * Huan(202111): We postpone the `this._currentUserId = undefined` to here,
             *  in case of the `logout` event listener need to check the `this.currentUserId`
             */
            await new Promise(resolve => setImmediate(() => {
                this.__currentUserId = undefined;
                resolve();
            }));
        }
        /**
         * @deprecated use `currentUserId` instead. (will be removed after Dec 31, 2022)
         */
        selfId() {
            config_js_1.log.warn('PuppetLoginMixin', 'selfId() is deprecated, use `currentUserId` instead:\n%s', new Error().stack);
            return this.currentUserId;
        }
        /**
         * @deprecated use isLoggedIn instead. will be removed after Dec 31, 2022
         */
        logonoff() {
            config_js_1.log.warn('PuppetLoginMixin', 'logonoff() is deprecated, use `isLoggedIn` instead:\n%s', new Error().stack);
            return this.isLoggedIn;
        }
    }
    return LoginMixin;
};
exports.loginMixin = loginMixin;
