// cmd-ts/dist/esm/positional.js
import * as Result from './Result';
import { string } from './types';
import chalk from 'chalk';
function fullPositional(config) {
    var _a, _b, _c;
    const displayName = (_b = (_a = config.displayName) !== null && _a !== void 0 ? _a : config.type.displayName) !== null && _b !== void 0 ? _b : 'arg';
    return {
        description: (_c = config.description) !== null && _c !== void 0 ? _c : config.type.description,
        helpTopics() {
            var _a, _b, _c, _d;
            const defaults = [];
            const defaultValueFn = (_a = config.defaultValue) !== null && _a !== void 0 ? _a : config.type.defaultValue;
            if (defaultValueFn) {
                try {
                    const defaultValue = defaultValueFn();
                    if ((_b = config.defaultValueIsSerializable) !== null && _b !== void 0 ? _b : config.type.defaultValueIsSerializable) {
                        defaults.push('default: ' + chalk.italic(defaultValue));
                    }
                    else {
                        defaults.push('optional');
                    }
                }
                catch (e) { }
            }
            const usage = defaults.length > 0 ? `[${displayName}]` : `<${displayName}>`;
            return [
                {
                    category: 'arguments',
                    usage,
                    description: (_d = (_c = config.description) !== null && _c !== void 0 ? _c : config.type.description) !== null && _d !== void 0 ? _d : 'self explanatory',
                    defaults,
                },
            ];
        },
        register(_opts) { },
        async parse({ nodes, visitedNodes, }) {
            var _a;
            const positionals = nodes.filter((node) => node.type === 'positionalArgument' && !visitedNodes.has(node));
            const defaultValueFn = (_a = config.defaultValue) !== null && _a !== void 0 ? _a : config.type.defaultValue;
            let positional = positionals[0];
            if (!positional) {
                if (defaultValueFn) {
                    return Result.ok(defaultValueFn());
                }
                else {
                    return Result.err({
                        errors: [
                            {
                                nodes: [],
                                message: `No value provided for ${displayName}`,
                            },
                        ],
                    });
                }
            }
            visitedNodes.add(positional);
            const decoded = await Result.safeAsync(config.type.from(positional.raw));
            if (Result.isErr(decoded)) {
                return Result.err({
                    errors: [
                        {
                            nodes: [positional],
                            message: decoded.error.message,
                        },
                    ],
                });
            }
            return Result.ok(decoded.value);
        },
    };
}
export function positional(config) {
    return fullPositional({
        type: string,
        ...config,
    });
}
