// wechaty-token/dist/cjs/src/cli/discover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discover = void 0;
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
const wechaty_token_js_1 = require("../wechaty-token.js");
async function handler(args) {
    let address;
    try {
        address = await new wechaty_token_js_1.WechatyToken(args.token).discover();
    }
    catch (e) {
        console.error(e);
        process.exit(1);
    }
    if (address) {
        /**
          * Huan(202108): `ip` is deprecated. use `host` instead
          *   See: https://github.com/wechaty/wechaty-puppet-service/issues/154
          */
        delete address['ip'];
        /**
         * Print the result as JSON.stringify format
         */
        console.info(JSON.stringify(address, null, '  '));
    }
    else {
        console.error([
            `NotFound: ${args.token}`,
            '',
            'Please contact your Wechaty Puppet Service Provider to get a valid token.',
            '',
            'Learn more from Wechaty official website at https://wechaty.js.org/docs/puppet-services/',
            '',
        ].join('\n'));
        process.exit(1);
    }
}
const discover = (0, cmd_ts_1.command)({
    name: 'discover',
    description: 'Wechaty TOKEN Service Discovery',
    args: {
        token: (0, cmd_ts_1.positional)({
            type: cmd_ts_1.string,
            description: 'Wechaty Puppet Service TOKEN',
        }),
    },
    handler,
});
exports.discover = discover;
