// wechaty-grpc/dist/cjs/src/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.Timestamp = exports.StringValue = exports.puppet = exports.proto = exports.openApi = exports.grpc = exports.google = exports.chunkDecoder = exports.chunkEncoder = void 0;
const grpc = __importStar(require("@grpc/grpc-js"));
exports.grpc = grpc;
const proto = __importStar(require("./proto.js"));
exports.proto = proto;
const openApi = __importStar(require("./openapi.js"));
exports.openApi = openApi;
const cjs_js_1 = require("./cjs.js");
Object.defineProperty(exports, "google", { enumerable: true, get: function () { return cjs_js_1.google; } });
Object.defineProperty(exports, "puppet", { enumerable: true, get: function () { return cjs_js_1.puppet; } });
const config_js_1 = require("./config.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_js_1.VERSION; } });
const google_js_1 = require("./google.js");
Object.defineProperty(exports, "StringValue", { enumerable: true, get: function () { return google_js_1.StringValue; } });
Object.defineProperty(exports, "Timestamp", { enumerable: true, get: function () { return google_js_1.Timestamp; } });
const chunk_transformer_js_1 = require("./chunk-transformer.js");
Object.defineProperty(exports, "chunkEncoder", { enumerable: true, get: function () { return chunk_transformer_js_1.chunkEncoder; } });
Object.defineProperty(exports, "chunkDecoder", { enumerable: true, get: function () { return chunk_transformer_js_1.chunkDecoder; } });
