// cmd-ts/dist/cjs/index.js
"use strict";
/**
 * The index module: the entrance to the world of cmd-ts 😎
 *
 * @packageDocumentation
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.oneOf = exports.union = exports.multioption = exports.multiflag = exports.restPositionals = exports.run = exports.runSafely = exports.dryRun = exports.positional = exports.option = exports.flag = exports.command = exports.binary = exports.extendType = exports.subcommands = void 0;
var subcommands_1 = require("./subcommands");
Object.defineProperty(exports, "subcommands", { enumerable: true, get: function () { return subcommands_1.subcommands; } });
var type_1 = require("./type");
Object.defineProperty(exports, "extendType", { enumerable: true, get: function () { return type_1.extendType; } });
__exportStar(require("./types"), exports);
var binary_1 = require("./binary");
Object.defineProperty(exports, "binary", { enumerable: true, get: function () { return binary_1.binary; } });
var command_1 = require("./command");
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return command_1.command; } });
var flag_1 = require("./flag");
Object.defineProperty(exports, "flag", { enumerable: true, get: function () { return flag_1.flag; } });
var option_1 = require("./option");
Object.defineProperty(exports, "option", { enumerable: true, get: function () { return option_1.option; } });
var positional_1 = require("./positional");
Object.defineProperty(exports, "positional", { enumerable: true, get: function () { return positional_1.positional; } });
var runner_1 = require("./runner");
Object.defineProperty(exports, "dryRun", { enumerable: true, get: function () { return runner_1.dryRun; } });
Object.defineProperty(exports, "runSafely", { enumerable: true, get: function () { return runner_1.runSafely; } });
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return runner_1.run; } });
var restPositionals_1 = require("./restPositionals");
Object.defineProperty(exports, "restPositionals", { enumerable: true, get: function () { return restPositionals_1.restPositionals; } });
var multiflag_1 = require("./multiflag");
Object.defineProperty(exports, "multiflag", { enumerable: true, get: function () { return multiflag_1.multiflag; } });
var multioption_1 = require("./multioption");
Object.defineProperty(exports, "multioption", { enumerable: true, get: function () { return multioption_1.multioption; } });
var union_1 = require("./union");
Object.defineProperty(exports, "union", { enumerable: true, get: function () { return union_1.union; } });
var oneOf_1 = require("./oneOf");
Object.defineProperty(exports, "oneOf", { enumerable: true, get: function () { return oneOf_1.oneOf; } });
