// wechaty-token/dist/src/cli/discover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discover = void 0;
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
const wechaty_token_1 = require("../wechaty-token");
async function handler(args) {
    let address;
    try {
        address = await new wechaty_token_1.WechatyToken(args.token).discover();
    }
    catch (e) {
        console.error(e);
        process.exit(1);
    }
    if (address) {
        /**
          * Huan(202108): `ip` is deprecated. use `host` instead
          *   See: https://github.com/wechaty/wechaty-puppet-service/issues/154
          */
        delete address['ip'];
        // print the result
        console.info(address);
    }
    else {
        console.info('NotFound');
        process.exit(1);
    }
}
const discover = cmd_ts_1.command({
    name: 'discover',
    description: 'Wechaty TOKEN Service Discovery',
    args: {
        token: cmd_ts_1.positional({
            type: cmd_ts_1.string,
            description: 'Wechaty Puppet Service TOKEN',
        }),
    },
    handler,
});
exports.discover = discover;
