// wechaty-puppet/dist/cjs/src/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.StateSwitch = exports.NAME = exports.MemoryCard = exports.log = exports.FileBox = exports.envVars = void 0;
const file_box_1 = require("file-box");
Object.defineProperty(exports, "FileBox", { enumerable: true, get: function () { return file_box_1.FileBox; } });
const brolog_1 = require("brolog");
const memory_card_1 = require("memory-card");
Object.defineProperty(exports, "MemoryCard", { enumerable: true, get: function () { return memory_card_1.MemoryCard; } });
const state_switch_1 = require("state-switch");
Object.defineProperty(exports, "StateSwitch", { enumerable: true, get: function () { return state_switch_1.StateSwitch; } });
const package_json_js_1 = require("./package-json.js");
const envVars = __importStar(require("./env-vars.js"));
exports.envVars = envVars;
const log = new brolog_1.Brolog();
exports.log = log;
const logLevel = process.env['WECHATY_LOG'];
if (logLevel) {
    log.level(logLevel.toLowerCase());
    log.silly('Puppet', 'Config: WECHATY_LOG set level to %s', logLevel);
}
const VERSION = package_json_js_1.packageJson.version || '0.0.0';
exports.VERSION = VERSION;
const NAME = package_json_js_1.packageJson.name || 'NONAME';
exports.NAME = NAME;
