// ducks/dist/cjs/src/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.validateDuck = exports.noopReducer = exports.Ducks = exports.Bundle = exports.Api = void 0;
/**
 *   Ducks - https://github.com/huan/ducks
 *
 *   @copyright 2020 Huan LI (李卓桓) <https://github.com/huan>
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
const config_js_1 = require("./config.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_js_1.VERSION; } });
const mod_js_1 = require("./ducks/mod.js");
Object.defineProperty(exports, "Ducks", { enumerable: true, get: function () { return mod_js_1.Ducks; } });
Object.defineProperty(exports, "noopReducer", { enumerable: true, get: function () { return mod_js_1.noopReducer; } });
const validate_duck_js_1 = require("./validate-duck.js");
Object.defineProperty(exports, "validateDuck", { enumerable: true, get: function () { return validate_duck_js_1.validateDuck; } });
const Api = __importStar(require("./api/mod.js"));
exports.Api = Api;
const bundle_js_1 = require("./bundle.js");
Object.defineProperty(exports, "Bundle", { enumerable: true, get: function () { return bundle_js_1.Bundle; } });
