// state-switch/dist/cjs/src/service-ctl-fsm/machine-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const inactive = {
    // entry : 'onEntry',
    // exit  : 'onExit',
    on: {
        START: 'starting',
    },
};
const starting = {
    invoke: {
        onDone: 'active',
        onError: 'canceled',
        src: 'start',
    },
    on: {
        CANCEL: 'inactive',
    },
};
const active = {
    // entry : 'onEntry',
    // exit  : 'onExit',
    on: {
        RESET: 'resetting',
        STOP: 'stopping',
    },
};
const stopping = {
    invoke: {
        onDone: 'inactive',
        onError: 'canceled',
        src: 'stop',
    },
    on: {
        CANCEL: 'inactive',
    },
};
const resetting = {
    invoke: {
        onDone: 'active',
        onError: 'canceled',
        src: 'reset',
    },
    on: {
        CANCEL: 'inactive',
    },
};
const canceled = {
    // entry : 'onEntry',
    // exit  : 'onExit',
    on: {
        RESET: 'resetting',
        START: 'starting',
        STOP: 'stopping',
    },
};
const states = {
    active,
    canceled,
    inactive,
    resetting,
    starting,
    stopping,
};
const initialContext = () => ({
    counter: 0,
});
const config = {
    context: initialContext(),
    id: 'ServiceCtlMachine',
    initial: 'inactive',
    states,
};
exports.config = config;
