// wechaty-redux/dist/cjs/src/duck/types.js
"use strict";
/**
 *   Wechaty Open Source Software - https://github.com/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NOP_COMMAND = exports.UNBIND_WECHATY_PUPPET_COMMAND = exports.BIND_WECHATY_PUPPET_COMMAND = exports.DEREGISTER_WECHATY_COMMAND = exports.REGISTER_WECHATY_COMMAND = exports.DEREGISTER_PUPPET_COMMAND = exports.REGISTER_PUPPET_COMMAND = exports.RESET_COMMAND_RESPONSE = exports.RESET_COMMAND = exports.DING_COMMAND_RESPONSE = exports.DING_COMMAND = exports.SCAN_RECEIVED_EVENT = exports.ROOM_TOPIC_RECEIVED_EVENT = exports.ROOM_LEAVE_RECEIVED_EVENT = exports.ROOM_JOIN_RECEIVED_EVENT = exports.ROOM_INVITE_RECEIVED_EVENT = exports.RESET_RECEIVED_EVENT = exports.READY_RECEIVED_EVENT = exports.MESSAGE_RECEIVED_EVENT = exports.LOGOUT_RECEIVED_EVENT = exports.LOGIN_RECEIVED_EVENT = exports.HEARTBEAT_RECEIVED_EVENT = exports.FRIENDSHIP_RECEIVED_EVENT = exports.ERROR_RECEIVED_EVENT = exports.DONG_RECEIVED_EVENT = exports.STATE_INACTIVATED_EVENT = exports.STATE_ACTIVATED_EVENT = exports.STOPPED_EVENT = exports.STARTED_EVENT = void 0;
/**
 * Pure Events
 */
exports.STARTED_EVENT = 'wechaty-redux/STARTED_EVENT';
exports.STOPPED_EVENT = 'wechaty-redux/STOPPED_EVENT';
exports.STATE_ACTIVATED_EVENT = 'wechaty-redux/STATE_ACTIVATED_EVENT';
exports.STATE_INACTIVATED_EVENT = 'wechaty-redux/STATE_INACTIVATED_EVENT';
exports.DONG_RECEIVED_EVENT = 'wechaty-redux/DONG_RECEIVED_EVENT';
exports.ERROR_RECEIVED_EVENT = 'wechaty-redux/ERROR_RECEIVED_EVENT';
exports.FRIENDSHIP_RECEIVED_EVENT = 'wechaty-redux/FRIENDSHIP_RECEIVED_EVENT';
exports.HEARTBEAT_RECEIVED_EVENT = 'wechaty-redux/HEARTBEAT_RECEIVED_EVENT';
exports.LOGIN_RECEIVED_EVENT = 'wechaty-redux/LOGIN_RECEIVED_EVENT';
exports.LOGOUT_RECEIVED_EVENT = 'wechaty-redux/LOGOUT_RECEIVED_EVENT';
exports.MESSAGE_RECEIVED_EVENT = 'wechaty-redux/MESSAGE_RECEIVED_EVENT';
exports.READY_RECEIVED_EVENT = 'wechaty-redux/READY_RECEIVED_EVENT';
exports.RESET_RECEIVED_EVENT = 'wechaty-redux/RESET_RECEIVED_EVENT';
exports.ROOM_INVITE_RECEIVED_EVENT = 'wechaty-redux/ROOM_INVITE_RECEIVED_EVENT';
exports.ROOM_JOIN_RECEIVED_EVENT = 'wechaty-redux/ROOM_JOIN_RECEIVED_EVENT';
exports.ROOM_LEAVE_RECEIVED_EVENT = 'wechaty-redux/ROOM_LEAVE_RECEIVED_EVENT';
exports.ROOM_TOPIC_RECEIVED_EVENT = 'wechaty-redux/ROOM_TOPIC_RECEIVED_EVENT';
exports.SCAN_RECEIVED_EVENT = 'wechaty-redux/SCAN_RECEIVED_EVENT';
/**
* Commands & Responses - Wechaty Puppet
*/
exports.DING_COMMAND = 'wechaty-redux/DING_COMMAND';
exports.DING_COMMAND_RESPONSE = 'wechaty-redux/DING_COMMAND_RESPONSE';
exports.RESET_COMMAND = 'wechaty-redux/RESET_COMMAND';
exports.RESET_COMMAND_RESPONSE = 'wechaty-redux/RESET_COMMAND_RESPONSE';
/**
 * @private Commands - Redux
 *
 * Wechaty Redux provide the minimum commands for the system state management:
 *  1. ding: request the `dong` event
 *  2. reset: reset the Wechaty state
 *
 * For more commands, please refer to the Wechaty CQRS:
 *  @link https://github.com/wechaty/cqrs
 */
exports.REGISTER_PUPPET_COMMAND = 'wechaty-redux/REGISTER_PUPPET_COMMAND';
exports.DEREGISTER_PUPPET_COMMAND = 'wechaty-redux/DEREGISTER_PUPPET_COMMAND';
exports.REGISTER_WECHATY_COMMAND = 'wechaty-redux/REGISTER_WECHATY_COMMAND';
exports.DEREGISTER_WECHATY_COMMAND = 'wechaty-redux/DEREGISTER_WECHATY_COMMAND';
exports.BIND_WECHATY_PUPPET_COMMAND = 'wechaty-redux/BIND_WECHATY_PUPPET_COMMAND';
exports.UNBIND_WECHATY_PUPPET_COMMAND = 'wechaty-redux/UNBIND_WECHATY_PUPPET_COMMAND';
exports.NOP_COMMAND = 'wechaty-redux/NOP_COMMAND';
