// wechaty-puppet/dist/esm/src/schemas/friendship.js
export var FriendshipType;
(function (FriendshipType) {
    FriendshipType[FriendshipType["Unknown"] = 0] = "Unknown";
    FriendshipType[FriendshipType["Confirm"] = 1] = "Confirm";
    FriendshipType[FriendshipType["Receive"] = 2] = "Receive";
    FriendshipType[FriendshipType["Verify"] = 3] = "Verify";
})(FriendshipType || (FriendshipType = {}));
/**
 * Huan(202002): Does those numbers are the underlying Wechat Protocol Data Values?
 */
export var FriendshipSceneType;
(function (FriendshipSceneType) {
    FriendshipSceneType[FriendshipSceneType["Unknown"] = 0] = "Unknown";
    FriendshipSceneType[FriendshipSceneType["QQ"] = 1] = "QQ";
    FriendshipSceneType[FriendshipSceneType["Email"] = 2] = "Email";
    FriendshipSceneType[FriendshipSceneType["Weixin"] = 3] = "Weixin";
    FriendshipSceneType[FriendshipSceneType["QQtbd"] = 12] = "QQtbd";
    FriendshipSceneType[FriendshipSceneType["Room"] = 14] = "Room";
    FriendshipSceneType[FriendshipSceneType["Phone"] = 15] = "Phone";
    FriendshipSceneType[FriendshipSceneType["Card"] = 17] = "Card";
    FriendshipSceneType[FriendshipSceneType["Location"] = 18] = "Location";
    FriendshipSceneType[FriendshipSceneType["Bottle"] = 25] = "Bottle";
    FriendshipSceneType[FriendshipSceneType["Shaking"] = 29] = "Shaking";
    FriendshipSceneType[FriendshipSceneType["QRCode"] = 30] = "QRCode";
})(FriendshipSceneType || (FriendshipSceneType = {}));
